### ISSUE TITLE
Make Laplace transform work on undefined functions

### ISSUE THREAD

**Issue by sympy-issue-migrator** [sympy/sympy#7219]
Example:
```
In [2]: x = Function('x')

In [3]: from sympy.abc import s

In [4]: laplace_transform(x(t).diff(), t, s)
Out[4]: LaplaceTransform(Derivative(x(t), t), t, s)
```
I expected `s * x(s)`.

The inverse doesn't work either:
```
In [5]: inverse_laplace_transform(s * x(s), s, t)
Out[5]: InverseLaplaceTransform(s*x(s), s, t, _None)
```
Aaron Meurer suggested either to implement a table lookup for some simple rules and properties or to just make the Integral work.

Original issue for #7219: http://code.google.com/p/sympy/issues/detail?id=4120
Original author: https://code.google.com/u/104623634006529252977/

**Comment by eric-wieser** [comment]
Here's a crappy implementation that catches the simple cases correctly (ignoring initial conditions):

```python
def laplace(e, t, s):
    res = laplace_transform(e, t, s, noconds=True)
    wf = Wild('f')
    lw = LaplaceTransform(wf, t, s)

    
    for exp in res.find(lw):
        e = exp.match(lw)[wf]
        
        if not isinstance(e, Derivative):
            continue
            
        if all(arg == t for arg in e.args[1:]):
            newexp = s**len(e.args[1:]) * LaplaceTransform(e.args[0], t, s)
            res = res.replace(exp, newexp)
            
    return res
```

**Comment by awa5114** [comment]
Has this been fixed yet?

**Comment by physici** [comment]
Any news here? Have the same problem that I need to calculate the Laplace transformation of an unknown function.

**Comment by adeak** [comment]
I updated Eric's ["crappy implementation"](https://github.com/sympy/sympy/issues/7219#issuecomment-154768904) for the corresponding question on Stack Overflow. It seems to me that the representation of `Derivative`s has changed, and that we can shoehorn in a hack for `Integral`s similarly. Here's my updated workaround for completeness' sake:
```
from sympy import *

def laplace(e, t, s):
    res = laplace_transform(e, t, s, noconds=True)
    wf = Wild('f')
    lw = LaplaceTransform(wf, t, s)

    for exp in res.find(lw):
        e = exp.match(lw)[wf]
        args = e.args

        if isinstance(e, Derivative):
            # for derivative check that there's only d/dt^n with n>0
            if len(args) == 2 and args[1][0] == t:
                n = args[1][1]
                if n > 0:
                    newexp = s**n * LaplaceTransform(e.args[0], t, s)
                res = res.replace(exp, newexp)

        elif isinstance(e, Integral):
            # for integral check that there's only n consecutive indefinite integrals w.r.t. t
            if all(len(arg) == 1 and arg[0] == t for arg in args[1:]):
                newexp = s**(-len(args[1:])) * LaplaceTransform(args[0], t, s)
                res = res.replace(exp, newexp)

        # otherwise don't do anything

    return res
```

I tested it with
```
x = Function('x')
s,t = symbols('s t')
print(laplace(Derivative(x(t), t, 3), t, s))
print(laplace(Integral(Integral(x(t), t), t), t, s))
```
and the results seem to make sense.

**Comment by freckletonj** [comment]
I ran into this same problem, and for future reference this is what @adeak 's solution `print`s:


```
import sympy as sym

x = sym.Function('x')
s,t = sym.symbols('s t')
a = sym.Derivative(x(t), t, 3)
b = sym.Integral(sym.Integral(x(t), t), t)

print()
print(laplace(a, t, s))
print(laplace(b, t, s))

print()
print(sym.laplace_transform(a, t, s))
print(sym.laplace_transform(b, t, s))

#####

s**3*LaplaceTransform(x(t), t, s)
LaplaceTransform(x(t), t, s)/s**2

LaplaceTransform(Derivative(x(t), (t, 3)), t, s)
LaplaceTransform(Integral(x(t), t, t), t, s)
```

**Comment by MrSamuelLaw** [comment]
Hi, has a workable solution been implemented as of 2021? Sympy has basically allowed me to skate through all the problems with Matlab up to this point and now it can't even do a simple transform for an undefined function? This seems like a gaping hole in an otherwise flawless package.

**Comment by oscarbenjamin** [comment]
Is this not just a problem with the way that the transform is displayed by the printer?

See https://github.com/sympy/sympy/issues/17098

**Comment by adeak** [comment]
@hanspi42 would that not be too specific? I imagine functions could then similarly be aware of their derivative, their integral, and all other kinds of transforms.

**Comment by hanspi42** [comment]
> 
> 
> @hanspi42 would that not be too specific? I imagine functions could then similarly be aware of their derivative, their integral, and all other kinds of transforms.

Yes, you are right. What about writing a new `laplace_transfrorm_functions` that has a dictionary of transform pairs as an additional argument? That dictionary could also contain initial conditions and thus make transforming ODEs possible.

**Comment by hanspi42** [comment]
Since I already did the specific code for the delta functions, I'd like do do this too, is it OK if I take this task?

The rules I would implement are:

[laplace_rules.pdf](https://github.com/sympy/sympy/files/7359558/laplace_rules.pdf)

Are there more rules you would think would be goo, or is this OK for the start?

I would of course also modify `inverse_laplace_transform` in the same way.

**Comment by hanspi42** [comment]
> That seems like a good start. There are more rules here that could also be added although it doesn't all have to be done in one go: https://en.wikipedia.org/wiki/List_of_Laplace_transforms

Ah, yes, those I would call "transform pairs" and I have a muuuuuuuch longer list than that one ;)))

Anyway, I'd suggest

1. build in the generic rules I have mentioned, starting with linearity,
2. build initial conditions into the diff rule,
3. do the same in  the inverse transform,
4. then check what it cannot yet do and build in more knowledge step by step as needed.

It can already do surprisingly much by integrating.

**Comment by hanspi42** [comment]
I think the linearity rule is best implemented as I already did it here:

https://github.com/sympy/sympy/blob/bf6345e907583199c1affcdd4356628239868624/sympy/integrals/transforms.py#L1278-L1284

For the rest, I agree and will try to implement it in the following days/weeks.

**Comment by oscarbenjamin** [comment]
Are there any situations where you need to be careful about applying linearity because not all of the integrals converge? For example:
```python
In [10]: integrate(1/(x+2) - 1/(x+1), (x, 0, oo))
Out[10]: -log(2)

In [11]: integrate(1/(x+2), (x, 0, oo)) - integrate(1/(x+1), (x, 0, oo))
Out[11]: nan
```

**Comment by hanspi42** [comment]
> Now I need to find out why the variable in the conditions is suddenly `_s` ...

Found it; `meijerint` calls `s = _dummy('s', 'rewrite-single', f)` which generates a `_s` behind the scenes; I just don't understand yet why this remains in the final expressions.

**Comment by oscarbenjamin** [comment]
The `laplace_transform` function is just a function so you can add arguments to it. Or are you actually asking about `LaplaceTransform`? Is it necessary to change LaplaceTransform?

**Comment by oscarbenjamin** [comment]
You could give an argument to `LaplaceTransform.doit()` and then pas it through.

Does it actually need to be passed through though? Is it more complicated than just adding
```
.subs(LaplaceTransform(x(t), t, s), X(s))
```
at the end?

**Comment by oscarbenjamin** [comment]
I would have thought that `LaplaceTransform(x(t-a), t, s).doit()` would return `LaplaceTransform(x(t), t, s)*exp(-a*s)` and then `laplace_transform` can just substitute that at the end if `transfuncs` is given to get `X(s)*exp(-a*s)`.

Or am I missing something?

**Comment by hanspi42** [comment]
The solution I implemented is to first do the Laplace transform and then tell sympy the name of the function in the s domain as well as the initial conditions.  One example is one of the test cases:

https://github.com/sympy/sympy/blob/b5191b805eba30488a54cc0e46dbab66052a0429/sympy/utilities/tests/test_wester.py#L2936-L2957

If you would like to review, have a look at #22376 :)

**Comment by GustavoExel** [comment]
Here's @eric-wieser 's implementation but up to date with sympy 1.7.1, and substituiting initial conditions

```python
def laplace(e, t, s, ics=None):
    res = smp.laplace_transform(e, t, s, noconds=True)
    wf = smp.Wild('f')
    lw = smp.LaplaceTransform(wf, t, s)

    for exp in res.find(lw):
        e = exp.match(lw)[wf]

        if not isinstance(e, smp.Derivative):
            continue

        if t == e.args[1][0]:
            n = e.args[1][1]
            f = e.args[0]
            F = smp.LaplaceTransform(y, t, s)
            newexp = s**n*F
            if ics is not None:
                newexp -= sum(s**(n-i-1)*smp.diff(f,(t,i)) for i in range(n)).xreplace(conds)
            res = res.replace(exp, newexp)

    return res

**Comment by trgoodale** [comment]
That seems to work, except the line

`newexp = s**n*smp.LaplaceTransform(y, t, s)`

ought to read

`newexp = s**n*smp.LaplaceTransform(e.args[0], t, s)`

to give the correct name of the function inside the unevaluated LaplaceTransform.

**Comment by oscarbenjamin** [comment]
Now on master (or in SymPy 1.10) `laplace_transform` can handle derivatives:
```python
In [8]: x = Function('x')

In [9]: s, t = symbols('s, t')

In [10]: ics = {x(0): 10}

In [11]: laplace_transform(x(t).diff(t), t, s)
Out[11]: s⋅LaplaceTransform(x(t), t, s) - x(0)

In [12]: _.subs(ics)
Out[12]: s⋅LaplaceTransform(x(t), t, s) - 10
```
Is there something done by the recipes above that isn't handled already?

**Comment by trgoodale** [comment]
Sorry, I had thought I had the latest version of Sympy, and was googling for how to do derivatives and found this thread.  I've now seen that Sympy 1.10 is out and solves the problem.  Thanks.

**Comment by nickbrady** [comment]
Is it possible (or reasonable) to have the `laplace_transform` function work for partial derivatives as well?
For instance if I slightly change the conditions from OscarBenjamin's post to something often encountered in physical situations:

``` python
c = sp.Function('c')
x = sp.symbols('x')
s, t = sp.symbols('s, t')

initial_condition = {c(0,x): 0}

sp.laplace_transform(c(t,x).diff(t), t, s)
```
the function does not recognize that L[∂/∂t c(t,x)] --> s L[c(t,x)] - c(0,x)

**Comment by oscarbenjamin** [comment]
I think that seems reasonable. It's just a question of making it happen.

**Comment by nickbrady** [comment]
> 

Not sure how to make it happen, but [adeak](https://github.com/adeak)'s method seems to be able to easily handle partial derivatives.
I also edited it a bit so that the Laplace transform moves through derivatives and integrals *not* wrt `t`, which I think is useful when wanting to make variable substitution for L[c(t,x)].

``` python
def laplace(expr, t, s):
    """
    Transform an expression to the Laplace transform.
    """
    
    try:
        res = sp.laplace_transform(expr, t, s, noconds=True)
    except NotImplementedError:
        res = sp.LaplaceTransform(expr, t, s)
        
    wf = sp.Wild('f')
    lw = sp.LaplaceTransform(wf, t, s)

    for exp in res.find(lw):
        e = exp.match(lw)[wf]
        args = e.args

        if isinstance(e, sp.Derivative):
            # example: Derivative(c(t, x, z), (t, 3), (x, 5), (z, 2))
            t_arg = [arg for arg in args[1:] if arg[0] == t]   # find t in the derivative arguments
            _args = [arg for arg in args[1:] if arg[0] != t]   # all other arguments

            n = 0
            if t_arg:
                n = t_arg[0][1]     # d/d_^n with n>0
            if n > 0:
                newexp = s**n * sp.LaplaceTransform(args[0], t, s)
            if _args:   # Laplace transform passes through derivatives w.r.t. variables != t
                newexp = s**n * sp.Derivative(sp.LaplaceTransform(args[0], t, s), *_args)
            
            res = res.replace(exp, newexp)

        elif isinstance(e, sp.Integral):
            # example: Integral(c(t, x), (x, ll, ul), (t, ll, ul))
            t_arg = [arg for arg in args[1:] if arg[0] == t]   # find t's in the integral arguments
            _args = [arg for arg in args[1:] if arg[0] != t]   # all other arguments
            
            n = len(t_arg)
            if _args:
                newexp = s**(-n) * sp.Integral(sp.LaplaceTransform(args[0], t, s), *_args)
            else:   # Laplace transform passes through integrals w.r.t. variables != t
                newexp = s**(-n) * sp.LaplaceTransform(args[0], t, s)
                
            res = res.replace(exp, newexp)
    
    return res
```

tested it on a few functions:
``` python
import sympy as sp
t, s, x, z = sp.symbols('t s x z')
c = sp.Function('c')
conc = c(t, x, z)

d1 = conc.diff(t, 2)
d2 = conc.diff(x, 2)
d3 = conc.diff(z, 2).diff(t,3).diff(x,5)

display(laplace(d1, t, s))
display(laplace(d2, t, s))
display(laplace(d3, t, s))

i1 = sp.Integral(sp.Integral(conc, t), t)
i2 = sp.Integral(sp.Integral(conc, x), t)

display(laplace(i1, t, s))
display(laplace(i2, t, s))
```

**Comment by nickbrady** [comment]
Actually, using dummy function that converts `f(t, x1, x2, ...)` to `F(t)`, you can combine adeak's function with the `sp.laplace_transform` function. Just need to change `F(t)` back to `f(t, x1, x2, ...)`.

``` python
import sympy as sp
from sympy import Derivative as δ
sp.init_printing(use_latex='mathjax')

def laplace(expr, t, s):
    """
        Transform an expression to the Laplace transform.
    """
    
    try:
        res = sp.laplace_transform(expr, t, s, noconds=True)
        # res = sp.laplace_transform(expr, t, s)
    except NotImplementedError:
        res = sp.LaplaceTransform(expr, t, s)
        
    wf = sp.Wild('f')                       # Wildcard for the function
    lw = sp.LaplaceTransform(wf, t, s)      # laplace of the wildcard
    _F_ = sp.Function('F')                  # dummy function to convert f(t, x1, x2...) --> F(t)

    for exp in res.find(lw):
        e = exp.match(lw)[wf]
        args = e.args
        
        func = e.args[0]                    # d(func)/d()
        if t in func.args:
            e = e.replace(func, _F_(t))     # f(t, x1, x2...) --> F(t)
            args = e.args                   # redefine args
            func_t0 = func.subs(t, 0)       # f(0, x1, x2...) , at t=0

        if isinstance(e, sp.Derivative):
            # example: Derivative(c(t, x, z), (t, 3), (x, 5), (z, 2))
            t_arg = [arg for arg in args[1:] if arg[0] == t]   # find t in the derivative arguments
            _args = [arg for arg in args[1:] if arg[0] != t]   # all other arguments

            n = 0
            if t_arg:
                n = t_arg[0][1]     # d/d_^n with n>0
                e_new = sp.Derivative(args[0], *t_arg)
            else:
                e_new = args[0]
                
            if n > 0:
                newexp = sp.laplace_transform(e_new, t, s)
            if _args:   # Laplace transform passes through derivatives wrt variables != t
                newexp = sp.Derivative(sp.laplace_transform(e_new, t, s), *_args)
            
            res = res.replace(exp, newexp)
            res = res.replace(_F_(t), func)     # F(t) --> f(t, x1, x2...)
            res = res.replace(_F_(0), func_t0)  # F(0) --> f(0, x1, x2...)
            

        elif isinstance(e, sp.Integral):
            # example: Integral(c(t, x), (x, ll, ul), (t, ll, ul))
            t_arg = [arg for arg in args[1:] if arg[0] == t]   # find t's in the integral arguments
            _args = [arg for arg in args[1:] if arg[0] != t]   # all other arguments
            
            n = len(t_arg)
            if _args:
                newexp = s**(-n) * sp.Integral(sp.LaplaceTransform(args[0], t, s), *_args)
            else:   # Laplace transform passes through integral wrt variables != t
                newexp = s**(-n) * sp.LaplaceTransform(args[0], t, s)
                
            res = res.replace(exp, newexp)
            res = res.replace(_F_(t), func)     # F(t) --> f(t, x1, x2...)
            res = res.replace(_F_(0), func_t0)  # F(0) --> f(0, x1, x2...)
    
    return res
```

tested using
```
t, s, x, z = sp.symbols('t s x z')
c = sp.Function('c')
conc = c(t, x, z)

d1 = conc.diff(t, 2)
d2 = conc.diff(x, 2)
d3 = conc.diff(z, 2).diff(t,3).diff(x,5)

display(laplace(d1, t, s))
display(laplace(d2, t, s))
display(laplace(d3, t, s))
```

**Comment by Compizfox** [comment]
I think Sympy is handling the Laplace transform of expressions containing a constant times a derivative wrong.

The Laplace transform of the a derivative of a function itself is handled correctly:

```
>>> from sympy import *
>>> t, s = symbols('t, s')
>>> y = Function('y')(t)
>>> laplace_transform(Derivative(y, t), t, s)
s*LaplaceTransform(y(t), t, s) - y(0)
```

But, when asking for the Laplace transform of this derivative multiplied by a constant, it returns the exact same answer as before, i.e. without the constant:

```
>>> a = symbols('a')
>>> laplace_transform(a*Derivative(y, t), t, s)
s*LaplaceTransform(y(t), t, s) - y(0)
```

The correct Laplace transform of this should be `a*(s*LaplaceTransform(y(t), t, s) - y(0))`, not?

I'm on Sympy 1.10.1:

```
>>> sympy.__version__
'1.10.1'
```

**Comment by hanspi42** [comment]
I am very confused now.  There is a test case for this,

https://github.com/sympy/sympy/blob/39de9a2698ad4bb90681c0fdb70b30a78233145f/sympy/integrals/tests/test_transforms.py#L702-L704

but when I do 

```
f = Function("f")
LT = laplace_transform
print(LT(10*diff(f(t), (t, 1)), t, s))
```

in a Jupyter notebook, I get 

`s*LaplaceTransform(f(t), t, s) - f(0)`

I

**Comment by adeak** [comment]
@hanspi42 any chance that `@slow` tests are inadvertently skipped sometimes? Can you manually run that test to see if it runs?

**Comment by oscarbenjamin** [comment]
The constant multiplier case was fixed in #23308 but there has not been a release of SymPy since then. You should test with the SymPy master branch (and the 1.11 release branch).

**Comment by hanspi42** [comment]
Partial derivatives are near the top of my work plan now.

**Comment by nwsteg** [comment]
Hello. Potential bug, or I'm doing something incorrectly?

Target function: f(t)=-t*cos(5*t)

```python
import sympy as sp
t,s = sp.symbols('t s')
expr=-t*sp.cos(5*t)
result=sp.laplace_transform(expr,t,s)[0]
result
```
The output in JupyterLab is
![image](https://user-images.githubusercontent.com/75855220/217076025-0ef1d856-33f7-485a-9168-f33d72e02026.png)

But if I ask for the Laplace transform of the same function from WolframAlpha, I get 
![image](https://user-images.githubusercontent.com/75855220/217076198-bf4fb9d5-3b0d-474b-887f-6d146b72b60e.png)

It appears that sympy is missing a negative sign?

**Comment by adeak** [comment]
@nwsteg looks like this was a bug in the last release (1.11.1) but fixed in master.

1.11.1:
```py
>>> import sympy as sym
>>> sym.laplace_transform('t*cos(5*t)', 't', 's')
((s**2 - 25)/(s**4 + 50*s**2 + 625), 0, True)
>>> sym.laplace_transform('-t*cos(5*t)', 't', 's')
((s**2 - 25)/(s**4 + 50*s**2 + 625), 0, True)
```

master:
```py
>>> import sympy as sym
>>> sym.laplace_transform('t*cos(5*t)', 't', 's')
((s**2 - 25)/(s**4 + 50*s**2 + 625), 0, True)
>>> sym.laplace_transform('-t*cos(5*t)', 't', 's')
((25 - s**2)/(s**4 + 50*s**2 + 625), 0, True)
```

**Comment by hanspi42** [comment]
Yes, we fixed this in #24296
